### PR TITLE
make var_views assignable

### PR DESCRIPTION
**Pull request by t4c1**
## Summary
Makes it possible to assign to blocks, and other views into `var_value` (does not include assignments of single elements). 

## Tests
Tested assignment of one block to another that contains aliasing.

## Side Effects
With this there is no reason anymore to call `var_value` a static matrix :)

## Release notes

Blocks, and other views into `var_value` can now be assigned to.

## Checklist

- [x] Math issue #1805 

- [x] Copyright holder: Tadej Ciglarič

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

File: stan/math/rev/core/var.hpp
Comment (SteveBronder) [inline stan/math/rev/core/var.hpp:451]: Should we reset `this_vi->adj_` to `prev_val` here?

File: stan/math/rev/core/var.hpp
Comment (SteveBronder) [inline stan/math/rev/core/var.hpp:435]: docs

File: test/unit/math/rev/core/var_test.cpp
Comment (SteveBronder) [inline test/unit/math/rev/core/var_test.cpp:232]: Adding `stan::math::var_value<Eigen::MatrixXd> A_block = A_v.block(0, 0, 3, 3);` fails with
```
test/unit/math/rev/core/var_test.cpp: In member function ‘virtual void AgradRev_var_matrix_view_assignment_Test::TestBody()’:
test/unit/math/rev/core/var_test.cpp:228:61: error: conversion from ‘var_value<Eigen::Block<Eigen::Map<Eigen::Matrix<double, -1, -1>, 8, Eigen::Stride<0, 0> >, -1, -1, false>>’ to non-scalar type ‘var_value<Eigen::Matrix<double, -1, -1>>’ requested
  228 |   stan::math::var_value<Eigen::MatrixXd> A_block = A_v.block(0, 0, 3, 3);
```

I think we just need a little constructor to do that

```cpp
  template <typename S, require_convertible_t<S&, value_type>* = nullptr,
            require_any_not_plain_type_t<T, S>* = nullptr>
  var_value(const var_value<S>& other) : vi_(new vari_type(other.vi_->val_)) {
    reverse_pass_callback(
        [this_vi = this->vi_, other_vi = other.vi_]() mutable {
          this_vi->adj_ += other_vi->adj_;
        });
  }
```

**Review by SteveBronder — CHANGES_REQUESTED** [review]
Looks like I went a little coo-coo trying to be const correct here when it was not needed lol. Just a note to add a construct so 

```cpp
var_value<Eigen::MatrixXd> A_block = A_v.block(0, 0, 3, 3);
```

works and then this looks good! I think we will need that for the compiler unless we can get away with using auto in the generated code

it would also be good to add more tests, though I can add a bunch of stress tests in #2064

**Comment by SteveBronder** [comment]
Yo fyi I merged this into #2064 because I would need it there anyway. We can either do one branch at a time or do just #2064 your call

File: stan/math/rev/core/var.hpp
Comment (t4c1) [reply stan/math/rev/core/var.hpp:451]: No. First at this point in code `prev_val` contains the value of `this_vi->adj_`. From this reverse pass callback on `this_vi->adj_` should contain adjoint associated with the values in this block before they were overwritten. Since they were overwritten, they were not used in any other operations and their adjoint is 0, which is set on line 450.

File: stan/math/rev/core/var.hpp
Comment (t4c1) [reply stan/math/rev/core/var.hpp:435]: done

File: test/unit/math/rev/core/var_test.cpp
Comment (t4c1) [reply test/unit/math/rev/core/var_test.cpp:232]: Fixed and added a test.

**Comment by t4c1** [comment]
> it would also be good to add more tests, though I can add a bunch of stress tests in #2064

Yeah I tend to be a bit lazy with tests. I think this is enough to show it works. If you want to add more, feel free. Or if you want me to add something specific, feel free to ask.

I also intend to add something similar for scalars. However, I wan to avoid merge conflicts, so I will wait until #2064 is merged. Also it contains some changes I'd need to do anyway.

**Comment by t4c1** [comment]
@SteveBronder This is waiting for next review.

File: stan/math/rev/core/var.hpp
Comment (SteveBronder) [inline stan/math/rev/core/var.hpp:496]: Looking over all this I think we should go back to having another template parameter for `var_value` to do SFINAE on. eod we have a lot of stuff here that is just for eigen types and some stuff that's just for doubles so I think it would make more sense. There would be some code duplication but I think things will look a lot cleaner

```
template <typename T, typename = void>
class var_value;

template <typename T>
class var_value<T, require_floating_point_t<T>> {
  // double stuff  
}

template <typename T>
class var_value<T, require_eigen_t<T>> {
  // eigen stuff  
}
```

File: stan/math/rev/core/var.hpp
Comment (SteveBronder) [inline stan/math/rev/core/var.hpp:113]: Should we have each `operator=` have a constructor as well. If feels like it would be good for the API to have the same ones.

**Review by SteveBronder — APPROVED** [review]
So I think this looks good. Splitting out `var_value` can either be done in this PR or #2064

File: stan/math/rev/core/var.hpp
Comment (t4c1) [reply stan/math/rev/core/var.hpp:496]: I agree. Separate PR though.

File: stan/math/rev/core/var.hpp
Comment (t4c1) [reply stan/math/rev/core/var.hpp:113]: `require_all_plain_type_t<T, S>* = nullptr` case is handled by defailt copy constructor (T and S can in fact only be same type). We can not define copy assignment operator to do the same as that would also take non-plain types.

`require_any_not_plain_type_t<T, S>* = nullptr` case is handled by this constructor.

We need to explicitely define copy constructor just to delegate to one of the upper two cases, as copy constructor can not be templated and wiould not pich one of those two cases. So I think this is fine.

**Pull request merged**
